silo/gviews_v4.py

```diff
@@ -587,12 +585,57 @@ def get_sheets_from_google_spreadsheet(request):
 
 @login_required
 def oauth2callback(request):
-    if not xsrfutil.validate_token(settings.SECRET_KEY, str(request.GET['state']), request.user):
-        return  HttpResponseBadRequest()
+    if not xsrfutil.validate_token(settings.SECRET_KEY,
+                                   str(request.GET.get('state', '')),
+                                   request.user):
+        return HttpResponseBadRequest()
 
     flow = _get_oauth_flow()
     credential = flow.step2_exchange(request.GET)
     storage = Storage(GoogleCredentialsModel, 'id', request.user, 'credential')
     storage.put(credential)
     redirect_url = request.session['redirect_uri_after_step2']
     return HttpResponseRedirect(redirect_url)
+
+
+@login_required
+def store_oauth2_credential(request):
+    if request.method == 'GET':
+        return HttpResponseNotAllowed(['POST', 'OPTIONS'])
+
+    data = request.POST
+    flow = _get_oauth_flow()
+
+    if 'access_token' in data:
+        access_token = data['access_token']
+        if 'refresh_token' in data:
+            refresh_token = data['refresh_token']
+        else:
+            refresh_token = None
+            logger.info(
+                'Received token response with no refresh_token. Consider '
+                "reauthenticating with approval_prompt='force'.")
+        token_expiry = None
+        if 'expires_in' in data:
+            delta = datetime.timedelta(seconds=int(data['expires_in']))
+            token_expiry = delta + datetime.datetime.utcnow()
+
+        logger.info('Successfully retrieved access token')
+        credential = OAuth2Credentials(
+            access_token, flow.client_id, flow.client_secret,
+            refresh_token, token_expiry, flow.token_uri, flow.user_agent,
+            revoke_uri=flow.revoke_uri, id_token=None,
+            token_response=data, scopes=flow.scope,
+            token_info_uri=flow.token_info_uri)
+    else:
```

Comment:
This is not a blocker, but I would prefer to do at the beginning: `if 'access_token' not in data`. And then you don't need to introduce the normal flow above inside an `if` condition and then continue outside the if the line `storage = Storage(GoogleCredentialsModel, 'id', request.user, 'credential')`.